libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -2077,38 +2075,57 @@ string YulUtilFunctions::readFromStorageValueType(Type const& _type, size_t _off
 			_type.identifier();
 
 	return m_functionCollector.createFunction(functionName, [&] {
-		solAssert(_type.sizeOnStack() == 1, "");
-		return Whiskers(R"(
-			function <functionName>(slot) -> value {
-				value := <extract>(sload(slot))
+		Whiskers templ(R"(
+			function <functionName>(slot) -> <?split>addr, selector<!split>value</split> {
+				<?split>let</split> value := <extract>(sload(slot))
+				<?split>
+					addr, selector := <splitFunction>(value)
+				</split>
 			}
-		)")
-				("functionName", functionName)
-				("extract", extractFromStorageValue(_type, _offset, false))
-				.render();
+		)");
+		templ("functionName", functionName);
+		templ("extract", extractFromStorageValue(_type, _offset));
+		FunctionType const* funType = dynamic_cast<FunctionType const*>(&_type);
+		if (_splitFunctionTypes && funType && funType->kind() == FunctionType::Kind::External)
+		{
+			templ("split", true);
+			templ("splitFunction", splitExternalFunctionIdFunction());
+		}
+		else
+			templ("split", false);
+		return templ.render();
 	});
 }
+
 string YulUtilFunctions::readFromStorageValueTypeDynamic(Type const& _type, bool _splitFunctionTypes)
 {
 	solAssert(_type.isValueType(), "");
-	if (_type.category() == Type::Category::Function)
-		solUnimplementedAssert(!_splitFunctionTypes, "");
-
 	string functionName =
 		"read_from_storage_value_type_dynamic" +
 		string(_splitFunctionTypes ? "split_" : "") +
 		"_" +
 		_type.identifier();
 	return m_functionCollector.createFunction(functionName, [&] {
 		solAssert(_type.sizeOnStack() == 1, "");
-		return Whiskers(R"(
-			function <functionName>(slot, offset) -> value {
-				value := <extract>(sload(slot), offset)
+		Whiskers templ(R"(
+			function <functionName>(slot, offset) -> <?split>addr, selector<!split>value</split> {
+				<?split>let</split> value := <extract>(sload(slot), offset)
+				<?split>
+					addr, selector := <splitFunction>(value)
+				</split>
 			}
-		)")
-		("functionName", functionName)
-		("extract", extractFromStorageValueDynamic(_type, _splitFunctionTypes))
-		.render();
+		)");
+		templ("functionName", functionName);
+		templ("extract", extractFromStorageValueDynamic(_type));
+		FunctionType const* funType = dynamic_cast<FunctionType const*>(&_type);
+		if (_splitFunctionTypes && funType && funType->kind() == FunctionType::Kind::External)
+		{
+			templ("split", true);
+			templ("splitFunction", splitExternalFunctionIdFunction());
+		}
+		else
+			templ("split", false);
+		return templ.render();
```

Comment:
This is basically the same as the version above. Can we extract single function that will be used for both cases?

Reply:
Are you suggesting one yul function or one c++ function?

Reply:
One yul function, to be used as a return at these two places. What do you think?

Reply:
Turned it into a single C++ function.

Reply:
Looks better now